### PR TITLE
Draft of revised PLIC spec based on previous commit.

### PR DESCRIPTION
**Pull request by changab**
- In PLIC spec, I change the fixed values such as the bit mask of priority, length of the register, the base address, etc. to agnostic. This is more flexible and scalable for the different PLIC designs from IHV and fewer controversies of adopting SiFive PLIC spec as the industrial spec.

In PLIC spec, we can just standardize register blocks of Interrupt Priorities, Interrupt Pending Bits, Interrupt Enables, Priority Thresholds, Interrupt Claim Process and Interrupt Completion as the generic control register blocks for PLIC. IHV could add whatever add-on on their PLIC. We also standardize the operational parameters to configure each interrupt source in this spec.

- I put another section "RISC-V PLIC Specification Affinity" at the end of this document, which mentions the changes of ACPI MADT table for RISC-V PLIC. The purpose of putting ACPI change in PLIC spec is to give more information to reviewers about the intention and the direction of standardizing PLIC spec. 
I also added a TBD section for Device Tree syntax of PLIC. This may need someone to finalize it if all of you think it is necessary to be mentioned in here.

### CODE REVIEW

**Comment by fintelia** [comment]
I don't think your changes are compatible with SiFive's existing layout. Specifically, the hart priority threshold registers and claim/clear registers are interleaved for each context with reserved space to pad each context onto a separate 4KB page:
```
base + 0x200000: Priority threshold for context 0
base + 0x200004: Claim/complete for context 0
base + 0x200???: Reserved for context 0
base + 0x201000: Priority threshold for context 1
base + 0x201004: Claim/complete for context 1 
base + 0x201???: Reserved for context 1
base + 0x202000: Priority threshold for context 2
base + 0x202004: Claim/complete for context 2
base + 0x202???: Reserved for context 2
...
```
I'd also like to point out that making the device description so generic makes things considerably more annoying for software developers. For instance, instead of being able use `sw` instructions to access PLIC registers, the programmer would now have to switch over all possible register sizes (and deal with `sd` not being available on 32-bit processors, etc.)

**Comment by aswaterman** [comment]
Echoing @fintelia's concerns, "generic" is an anti-feature in this case.  The more flexible the interface is, the more complex a generic driver gets, and the more information it must depend on to do its job.  I don't think there's sufficient justification for that in this case.

**Comment by changab** [comment]
@fintelia I believe this PLIC spec is not trying to standardize every piece of SiFIve PLIC implementation as the industrial standard of RISC-V PLIC, however, SiFive PLIC is a good starting point and reference. The distributed register implementation is similar to SiFive one, just those address is different. I didn't follow register offset of SiFive PLIC, that is just an example. I can revise it if you think it's good for the example.

Thanks for the suggestion of x-bit reg thing. Although I put x-bit reg here, I also think 32-bit will be the real case of PLIC. 64-bit and 128-bit are too much. I will revise all x-bit to 32-bit.

I am not H/W guy, but there is no reason to restrict H/W vendors to implement PLIC follow whatever was done by someone before. The controversies may be something like is 1024 interrupt sources mandatory for any PLICs? Why reserve 4K space for only 53 interrupt sources? why 7-level interrupt priority in 3-bit takes whole 32-bit regs? Is a tiny PLIC really need to implement interrupt priority? such such...

The PLIC spec here should aim at standardizing PLIC SW/FW operation mechanism but not the H/W implementation IMO.  On SW/FW side, ACPI table accommodates any PLIC implementations follow this spec. One driver should be able to handle any PLIC implementations according to the information provided, so does Device Tree I guess. And we may not need to revisit some industrial standards (like ACPI) in the future if the original design is flexible enough.
@aswaterman I agree with your statement "The more flexible the interface is, the more complex a generic driver gets, and the more information it must depend on to do its job" though...

File: riscv-plic.adoc [hunk 1]
```diff
@@ -17,102 +17,436 @@ https://creativecommons.org/licenses/by/4.0/.
...
+
+#### Figure 5.24 PLIC-Global System Interrupts (Single Processor and Single PLIC Scenario)
+
+image::https://github.com/changab/riscv-plic-acpi-images/blob/master/Figure5-25.jpg[GitHub][1000,705]
+
+#### Figure 5.25 PLIC-Global System Interrupts (Multiple Processors and Multiple PLICs Scenario)
+
+
+## Device Tree Syntax for RISC-V PLIC
+*TBD*
```
Comment (fintelia) [inline riscv-plic.adoc]: For reference, the current device tree syntax is:
```
        interrupt-controller@c000000 {
            #interrupt-cells = <0x00000001>;
            compatible = "riscv,plic0";
            interrupt-controller;
            interrupts-extended = <0x00000010 0x0000000b 0x00000011 0x0000000b 0x00000011 0x00000009 0x00000012 0x0000000b 0x00000012 0x00000009 0x00000013 0x0000000b 0x00000013 0x00000009 0x00000014 0x0000000b 0x00000014 0x00000009>;
            reg = <0x00000000 0x0c000000 0x00000000 0x04000000>;
            reg-names = "control";
            riscv,max-priority = <0x00000007>;
            riscv,ndev = <0x00000035>;
            phandle = <0x0000000c>;
        };
```

Comment (changab) [reply riscv-plic.adoc]: Thanks, @fintelia
I just committed changes for Device Tree syntax

**Comment by dkruckemyer-ventana** [comment]
My understanding was that the SiFive PLIC was going to be the RISC-V PLIC for the purposes of creating the first version of the UNIX platform spec. From that perspective, I don't believe that any changes to the register format need to be made.

Moving forward, however, I think a PLIC 2.0 is needed to add support for more sophisticated interrupt distribution, local CSR interfaces, MSIs, virtualization, etc. For that spec, it's completely appropriate to rejig the register interface. Per @changab's point, we should focus on the SW interface for that rev so that a single driver can support multiple compliant HW implementations.

**Comment by changab** [comment]
@dkruckemyer-ventana Yes, we did not change register format. This change still based on SiFive PLIC and just gives more flexibilities to other PLIC designs. 

@all, I will wait for couple days and merge this PR to master if no one objects. We still can give comments on master.
thanks

**Comment by dkruckemyer-ventana** [comment]
I guess my comment about the register format really should have been more explicit to include the register address map, i.e. I didn't think anyone was signing up to change the PLIC 1.0 SW interface for the first version of the UNIX platform spec. Does anyone expect to implement the PLIC 1.0 "architecture" in a way that is different from the SiFive version? (I'm not really objecting, per se, just trying to square my understanding of the platform spec plan with these changes.)

**Comment by changab** [comment]
@dkruckemyer-ventana I understand your concern. However, it will be hard to make the fixed register format version into PC/Server industrial spec IMO. We have to come out the scalable and flexible spec of PLIC then submit ECR to PC/Server industrial spec such as UEFI, ACPI and etc..
Also, it always takes time to go through the engineer change review process in industrial spec working group. Frankly speaking, I would like to get it done at once rather than submit ECR to spec working group couple times for the different phases.
How about if I add another section says "The Recommendation of PLIC implementation" and adopt SiFive register format and memory map?

**Comment by changab** [comment]
This document is revised according to PLICv1 definitions. Please check if this version works for you.

File: riscv-plic.adoc [hunk 2]
```diff
@@ -17,102 +17,592 @@ https://creativecommons.org/licenses/by/4.0/.
...
+　　　　　　　　　riscv,ndev = <0x00000035>;			
+　　　　　　　　　riscv,int-context = <&cpu0 0x00000000 "M-Mode">;
+　　　　　　　　　riscv,int-context = <&cpu0 0x00000001 "M-Mode">; 
+　　　　　　　　　riscv,int-context = <&cpu0 0x00000001 "S-Mode">;
+　　　　　　　　　riscv,int-context = <&cpu0 0x00000002 "M-Mode">; 
+　　　　　　　　　riscv,int-context = <&cpu0 0x00000002 "S-Mode">; 
+　　　　　　　　　riscv,int-context = <&cpu0 0x00000003 "M-Mode">;
+　　　　　　　　　riscv,int-context = <&cpu0 0x00000003 "S-Mode">; 
+　　　　　　　　　riscv,int-context = <&cpu0 0x00000004 "M-Mode">; 
+　　　　　　　　　riscv,int-context = <&cpu0 0x00000004 "S-Mode">;
```
Comment (fintelia) [inline riscv-plic.adoc]: These lines are redundant to the `interupts-extended` attribute. Specifically, that attribute is formatted as pairs of cells where the first is the phandle of the hart (for which there is a mapping to hartid elsewhere in the device tree), and the second is the interrupt number (0xb=M-mode external interrupt, 0x9=S-mode external interrupt).

Comment (changab) [reply riscv-plic.adoc]: oh I see. I don't have ideas about what interrupt-extended means. thanks for reviewing this.

**Comment by changab** [comment]
Close this PR.

**Pull request closed without merging**